### PR TITLE
WIP certrotation: don't let multiple controllers rewrite metadata of the shared resource

### PR DESCRIPTION
**Pull request by vrutkovs**
This prevents http 409 errors when both labels and content need updating in one sync. See this [test failure](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-kube-apiserver-operator/1655/pull-ci-openshift-cluster-kube-apiserver-operator-master-e2e-aws-ovn/1767865041786245120) when new library-go is being pulled in cluster-kube-apiserver-operator

### CODE REVIEW

**Comment by vrutkovs** [comment]
/hold

Testing this in https://github.com/openshift/cluster-kube-apiserver-operator/pull/1655

**Comment by vrutkovs** [comment]
/hold cancel

Seems to work as expected

File: pkg/operator/certrotation/cabundle.go [hunk 1]
```diff
@@ -64,10 +64,11 @@ func (c CABundleConfigMap) EnsureConfigMapCABundle(ctx context.Context, signingC
 	}
 	needsMetadataUpdate = c.AdditionalAnnotations.EnsureTLSMetadataUpdate(&caBundleConfigMap.ObjectMeta) || needsMetadataUpdate
 	if needsMetadataUpdate && len(caBundleConfigMap.ResourceVersion) > 0 {
-		_, _, err := resourceapply.ApplyConfigMap(ctx, c.Client, c.EventRecorder, caBundleConfigMap)
+		actualCABundleConfigMap, _, err := resourceapply.ApplyConfigMap(ctx, c.Client, c.EventRecorder, caBundleConfigMap)
```
Comment (p0lyn0mial) [inline pkg/operator/certrotation/cabundle.go:67]: do you mind adding a unit test ?

**Comment by p0lyn0mial** [comment]
In the future we could refactor the controller to only issue a single update when meta and content require updating.

**Comment by p0lyn0mial** [comment]
@vrutkovs a wrong update ? I don't see a unit test.

File: pkg/operator/certrotation/cabundle.go (hunk 1 above)
Comment (vrutkovs) [reply pkg/operator/certrotation/cabundle.go:67]: Done, not sure if it really ensures hotlooping is not happening (imo fuzzing would confirm it)

Comment (p0lyn0mial) [reply pkg/operator/certrotation/cabundle.go:67]: This PR doesn't solve the hotloop issue. The hotloop issue was resolved by https://github.com/openshift/cluster-kube-apiserver-operator/pull/1661.

There were `4` controllers fighting over `kube-control-plane-signer-ca`

**Comment by p0lyn0mial** [comment]
> See this [test failure](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-kube-apiserver-operator/1655/pull-ci-openshift-cluster-kube-apiserver-operator-master-e2e-aws-ovn/1767865041786245120) when new library-go is being pulled in cluster-kube-apiserver-operator

For the record, the above failure was due to https://github.com/openshift/cluster-kube-apiserver-operator/pull/1661 not due to the issue addressed by this PR.

File: pkg/operator/certrotation/cabundle_test.go [hunk 2]
```diff
@@ -64,6 +63,66 @@ func TestEnsureConfigMapCABundle(t *testing.T) {
...
+				if !actions[0].Matches("get", "configmaps") {
+					t.Error(actions[0])
+				}
+				if !actions[1].Matches("update", "configmaps") {
+					t.Error(actions[1])
+				}
+				if !actions[2].Matches("get", "configmaps") {
+					t.Error(actions[2])
+				}
+				if !actions[3].Matches("update", "configmaps") {
```
Comment (p0lyn0mial) [inline pkg/operator/certrotation/cabundle_test.go:106]: the second call is unnecessary, we should refactor the controller to issue only a single API call.

Comment (p0lyn0mial) [reply pkg/operator/certrotation/cabundle_test.go:106]: in addition it is worth to also examine the payload of the request (updated configmap), please add it to your todo list.

**Comment by p0lyn0mial** [comment]
/assign @stlaz 

for approval

File: pkg/operator/certrotation/cabundle.go (hunk 1 above)
Comment (vrutkovs) [reply pkg/operator/certrotation/cabundle.go:75]: Fixed this by allowing metadata change to the first controller only (whoever has set the first ownership ref).

I'll run this a few time to make sure this won't make TLS metadata jitter from run to run

File: pkg/operator/certrotation/cabundle.go [hunk 3]
```diff
@@ -58,16 +58,25 @@ func (c CABundleConfigMap) EnsureConfigMapCABundle(ctx context.Context, signingC
 		)}
 	}
 
+	isPrimaryOwner := false
 	needsMetadataUpdate := false
 	if c.Owner != nil {
 		needsMetadataUpdate = ensureOwnerReference(&caBundleConfigMap.ObjectMeta, c.Owner)
+		if caBundleConfigMap.ObjectMeta.OwnerReferences[0] == *c.Owner {
```
Comment (p0lyn0mial) [inline pkg/operator/certrotation/cabundle.go:65]: Why can't we just have a single owner?

Comment (vrutkovs) [reply pkg/operator/certrotation/cabundle.go:65]: CA bundle can contain multiple signers, controlled by different controllers. We could have a single ownerRef, but that wouldn't reflect reality. It also makes easier to find such CA bundles in the must-gathers

Comment (p0lyn0mial) [reply pkg/operator/certrotation/cabundle.go:65]: > CA bundle can contain multiple signers, controlled by different controllers

This is wrong and should be fixed. Why can't we have a single signer ?
Could you point me to code that does this ?

Comment (p0lyn0mial) [reply pkg/operator/certrotation/cabundle.go:65]: Still, there is a single signer and a CA bundle. 
Also as far as I can tell `kas-o` doesn't make use of the `Owner` field for both signer and CA resources.

Comment (p0lyn0mial) [reply pkg/operator/certrotation/cabundle.go:65]: I don't understand why we have to have multiple controllers controlling a single resource. I would prefer to have a single controller for that.

It seems to me that we would like to mark which components are using the resource, which could be done with a single controller.

**Comment by vrutkovs** [comment]
Running one controller for a resource would fix this and few more problems, closing in favor of https://github.com/openshift/cluster-kube-apiserver-operator/pull/1669

**Pull request closed without merging**